test/test_encryption.py

```diff
@@ -1942,6 +1944,44 @@ def test_via_loading_shared_library(self):
         with self.assertRaises(ServerSelectionTimeoutError):
             no_mongocryptd_client.db.command("ping")
 
+    # https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/tests/README.rst#20-bypass-creating-mongocryptd-client-when-shared-library-is-loaded
+    @unittest.skipUnless(os.environ.get("TEST_CRYPT_SHARED"), "crypt_shared lib is not installed")
+    def test_client_via_loading_shared_library(self):
+        class TCPHandler(socketserver.BaseRequestHandler):
+            def handle(self):
+                assert not self.request.recv(1024)
+
+        def listen():
+            addr = ("localhost", 27021)
+            with socketserver.TCPServer(addr, TCPHandler) as sock:
+                sock.server_bind()
+                sock.server_activate()
```

Comment:
These two method are already called automatically. To actually start handling requests we need to: 
> call the [handle_request()](https://docs.python.org/3/library/socketserver.html#socketserver.BaseServer.handle_request) or [serve_forever()](https://docs.python.org/3/library/socketserver.html#socketserver.BaseServer.serve_forever) method of the server object to process one or many requests.

Reply:
Those two methods seem to not be called automatically--commenting out either one causes no error to be thrown in the thread. Also, I do not think we need to process any requests--we just need to confirm that the port is not in use.

Reply:
Ah, no, the issue is that those two methods are called automatically, but in a try block, so the error is suppressed. We need to call them ourselves to get the error.

Reply:
`handle_request` does not raise an error, and `serve_forever` ignores the timeout so it also does not raise an error and instead just blocks forever. The only functions that seem to work are the ones that I have.

Reply:
We can't call bind or server_activate here. Those methods are already called automatically. Calling sock.server_bind() here fails with errno 22 because you can't bind to the same address twice.

Reply:
We have to call either handle_request() or serve_forever() here or else the TCPServer will just immediately close and exit the with block.

Reply:
As I explained, they are called automatically *in a try block*, so the error produced is suppressed, thus we must do it ourselves to capture the exception. 
This same test ran with a port that is in use gives a different error, so I know that this is not just because of rebinding. Additionally, those two functions simply do not display the right behavior, in addition to being unnecessary--the connection code already tells us if the port is in use. I have tried the configuration you are suggesting and it just doesn't work.

Reply:
I believe you are conflating the behavior of TCP server and the TCP connections that the server spawns (the socketserver docs call these "requests" and each one gets handled with a new BaseRequestHandler instance). As written all this test does currently is assert that calling TCPServer.server_bind() twice raises an error:

```Python
>>> import socketserver
>>> with socketserver.TCPServer(("localhost", 47021), None) as sock:pass
... 
>>> with socketserver.TCPServer(("localhost", 47021), None) as sock:
...     sock.server_bind()
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/socketserver.py", line 466, in server_bind
    self.socket.bind(self.server_address)
OSError: [Errno 22] Invalid argument
```

Instead we need to call serve_forever() and implement a handler which causes the test to fail if a connection is established to the TCPServer. Here's an example:
```python
import socketserver
import threading
import time
from pymongo import MongoClient

# Setup server.
connection_established = [False]
class Handler(socketserver.BaseRequestHandler):
    def handle(self):
        connection_established[0] = True

server = socketserver.TCPServer(("localhost", 47021), Handler)
def serve():
    with server:
        server.serve_forever(poll_interval=.05)  # Short poll timeout to speed up the test

server_thread = threading.Thread(target=serve)
server_thread.start()
# Run test: connect=True will cause the test to fail, connect=False should pass
client = MongoClient("localhost", 47021, connect=False)
time.sleep(1)
# Shutdown server.
server.shutdown()
server_thread.join()
assert not connection_established[0], "a connection was established"
```

Reply:
Try your example with a port that is in use, though, and it gives a different error:
```
>>> with socketserver.TCPServer(("localhost", 27017), None) as sock:
...     sock.server_bind()
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/socketserver.py", line 452, in __init__
    self.server_bind()
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/socketserver.py", line 466, in server_bind
    self.socket.bind(self.server_address)
OSError: [Errno 48] Address already in use
```
Thus, as written this test does actually test that the port is not in use, that's why I added the assert at the end for the errno.

Reply:
We don't want to test a port that's already in use. The TCPServer has to bind to a free port because we're using it as a pretend mongocryptd server (which would also bind to a free port to listen for incoming connections). I suggest you do some reading on sockets, this page is a great guide: https://beej.us/guide/bgnet/html/split/system-calls-or-bust.html

In particular: https://beej.us/guide/bgnet/html/split/system-calls-or-bust.html#acceptthank-you-for-calling-port-3490

Note the page goes over C-level sockets but Python is very close in API.